File: Chapter 8 - Lecture/p18.m
% program18.m - Chebyshev differentiation via FFT
%               Compare program11.m
close all;
scrsz = get(groot,'ScreenSize');
figure('position', [100   100   4/5*scrsz(3)  4/5*scrsz(4)]); 

xx = linspace(-1,1,201);
ff = exp(xx).*sin(5*xx); clf
fs = 14;
for N = [10   20]
    
    % Plot the function.
    % x = cos(pi*(0:N)'/N);             % option 1
    x = cos(pi * linspace(0, 1, N)');   % option 2
    f = exp(x).*sin(5*x);
    subplot(2, 2, 2*(N == 20) + 1);
    plot(x, f,'.','markersize', 16), grid on
    line(xx,ff,'linewidth', 0.8, 'color', 'k');
    title(['f(x), N=' int2str(N)], 'FontSize', fs);
    
    
    % Plot the error of the derivatives.
    error = chebfft(f) - exp(x).*(sin(5*x)+5*cos(5*x));
    subplot(2, 2, 2*(N == 20) + 2);
    plot(x, error, '.', 'markersize', 16), grid on
    line(x, error, 'linewidth', 0.8, 'color', 'k');
    title(['error in f^{\prime}(x), N=', int2str(N)], 'FontSize', fs);
end

File: p18.py
# -*- coding: utf-8 -*-
"""
Chebyshev differentiation via FFT

@author: Khiem
"""

import numpy as np
import matplotlib.pyplot as plt
from cheb import chebfft

plt.close('all')
plt.figure(figsize=(7, 5))


xx = np.linspace(-1, 1, 201)
ff = np.exp(xx) * np.sin(5*xx)
for N in [18, 20]:
    x = np.cos(np.pi * np.arange(0,N+1)/N)
    f = np.exp(x) * np.sin(5*x)
    #plt.subplot(2,2,2*(N==20)+1)
    plt.subplot(position=[0.1, 0.6 - 0.5*(N==20), 0.35, 0.3])
    plt.plot(x, f, 'bo', xx, ff, 'k-', markersize=4, linewidth=1.2)
    plt.grid(True)
    plt.title(r'$f(x), N = ' + str(N) + '$', fontsize=12)
    
    error = chebfft(f) - np.exp(x) * (np.sin(5*x) + 5*np.cos(5*x))
    #plt.subplot(2,2,2*(N==20)+2)
    plt.subplot(position=[0.55, 0.6 - 0.5*(N==20), 0.35, 0.3])
    plt.plot(x, error, 'k-o', markersize=4, linewidth=1.2)
    plt.grid(True)
    plt.title(r"error in $f^{\prime}(x)$, $N =" + str(N) + "$", fontsize=12)
    

